src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,107 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int SIZE_ARRAY = 10;
+    private static final double GROW_FACTOR = 1.5;
+    private T[] array;
+    private int size;
+
+    public ArrayList() {
+        array = (T[])new Object[SIZE_ARRAY];
+    }
+
     @Override
     public void add(T value) {
-
+        growArray();
+        array[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index == size) {
+            add(value);
+            return;
+        }
+        checkIndexOutOfBoundsAdd(index);
+        if (size == array.length) {
+            growArray();
+        }
+        System.arraycopy(array, index, array, index + 1,size - index);
+        array[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size();i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexOutOfBounds(index);
+        return array[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndexOutOfBounds(index);
+        array[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexOutOfBounds(index);
+        T removed = array[index];
+        System.arraycopy(array, index + 1, array, index, size - index - 1);
+        size--;
+        return removed;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if ((element == null && array[i] == null)
```

Comment:
this statement may be simplified:
```suggestion
            if ((element == array[i])
```